Plugin fails to use remote value file when URL is composed of a complex query string

### Current Behavior

Hello,

We are storing our secret values in a file stored in a private Gitlab repository.

To leverage the download capacity of the plugin we set the following scheme : 

```
secrets+gpg-import:///secret.key?https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev
```

When templating a chart with this external value file, Helm fails with this error : 

```
helm template -f "secrets+gpg-import:///secret.key?https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev" . --debug
[...]
+++ _mktemp
+++ mktemp /tmp/tmp.gLZ3jvBu60/XXXXXX
++ _tmp_file='/tmp/tmp.gLZ3jvBu60/Q35zNo.yaml/raw?access_token=XXXX&ref=dev'
++ download 'https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev'
/home/antoine/.local/share/helm/plugins/helm-secrets/scripts/lib/file/http.sh: line 21: /tmp/tmp.gLZ3jvBu60/Q35zNo.yaml/raw?access_token=XXXXXX&ref=dev: No such file or directory
++ '[' false = true ']'
++ fatal 'Error while download url %s' 'https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev'
++ error 'Error while download url %s' 'https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev'
++ log 'Error while download url %s' 'https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev'
++ '[' 2 -le 1 ']'
++ format='Error while download url %s'
++ shift
++ printf '[helm-secrets] Error while download url %s\n' 'https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev'
[helm-secrets] Error while download url https://XXXX.gitlab.com/api/v4/projects/XXXX/repository/files/XXX%2Fgitops%2Fsecrets.yaml/raw?access_token=XXXXX&ref=dev
++ exit 1
[...]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x55568212840d]

goroutine 1 [running]:
panic(0x555682dcba80, 0x555683b23560)
        /usr/lib/golang/src/runtime/panic.go:1065 +0x565 fp=0xc00061f7c8 sp=0xc00061f700 pc=0x555680e04c05
runtime.panicmem()
        /usr/lib/golang/src/runtime/panic.go:212 +0x5d fp=0xc00061f7e8 sp=0xc00061f7c8 pc=0x555680e02bfd
runtime.sigpanic()
        /usr/lib/golang/src/runtime/signal_unix.go:734 +0x185 fp=0xc00061f820 sp=0xc00061f7e8 pc=0x555680e1be25
bytes.(*Buffer).Bytes(...)
        /usr/lib/golang/src/bytes/buffer.go:54
helm.sh/helm/v3/pkg/cli/values.readFile(0xc000315100, 0xf8, 0xc0002dbe80, 0x3, 0x4, 0x6, 0xc0002dbe80, 0x1, 0xc00061f950, 0x55568208f7ac)
        /builddir/build/BUILD/helm-3.5.4/_build/src/helm.sh/helm/v3/pkg/cli/values/options.go:120 +0x16d fp=0xc00061f890 sp=0xc00061f820 pc=0x55568212840d
helm.sh/helm/v3/pkg/cli/values.(*Options).MergeValues(0xc00010ef00, 0xc0002dbe80, 0x3, 0x4, 0x1, 0x26, 0x0)
        /builddir/build/BUILD/helm-3.5.4/_build/src/helm.sh/helm/v3/pkg/cli/values/options.go:48 +0x116 fp=0xc00061f960 sp=0xc00061f890 pc=0x5556821279b6
main.runInstall(0xc00054b8c0, 0x1, 0x4, 0xc000797a40, 0xc00010ef00, 0x555682ff1fa8, 0xc000010018, 0xc, 0x5556821c1c87, 0x4)
        /builddir/build/BUILD/helm-3.5.4/_build/src/helm.sh/helm/v3/cmd/helm/install.go:193 +0x251 fp=0xc00061fa10 sp=0xc00061f960 pc=0x55568219adf1
main.newTemplateCmd.func2(0xc00002d340, 0xc00054b8c0, 0x1, 0x4, 0x0, 0x0)
        /builddir/build/BUILD/helm-3.5.4/_build/src/helm.sh/helm/v3/cmd/helm/template.go:73 +0x136 fp=0xc00061fd38 sp=0xc00061fa10 pc=0x5556821bbcb6
github.com/spf13/cobra.(*Command).execute(0xc00002d340, 0xc00054b880, 0x4, 0x4, 0xc00002d340, 0xc00054b880)
        /usr/share/gocode/src/github.com/spf13/cobra/command.go:850 +0x472 fp=0xc00061fdf8 sp=0xc00061fd38 pc=0x555680f98332
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002e2dc0, 0xc000114398, 0x1, 0xc00061ff60)
        /usr/share/gocode/src/github.com/spf13/cobra/command.go:958 +0x375 fp=0xc00061fed8 sp=0xc00061fdf8 pc=0x555680f98eb5
github.com/spf13/cobra.(*Command).Execute(...)
        /usr/share/gocode/src/github.com/spf13/cobra/command.go:895
main.main()
        /builddir/build/BUILD/helm-3.5.4/_build/src/helm.sh/helm/v3/cmd/helm/helm.go:80 +0x25b fp=0xc00061ff88 sp=0xc00061fed8 pc=0x55568219897b
runtime.main()
        /usr/lib/golang/src/runtime/proc.go:225 +0x263 fp=0xc00061ffe0 sp=0xc00061ff88 pc=0x555680e07883
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc00061ffe8 sp=0xc00061ffe0 pc=0x555680e3bea1
```

### Expected Behavior

I would have expected the helm-secrets plugin to correctly fetch the remote value file and the templating to succeed.

### Steps To Reproduce

```markdown
1. Make a secret value file accessible from a complex URL with different query string arguments
2. Trigger the bug by templating an Helm chart using this remote secret values file as an input : the file download handled by the plugin should fail
```


### Environment

- Helm Version: `version.BuildInfo{Version:"v3.5", GitCommit:"", GitTreeState:"", GoVersion:"go1.16.6"}`
- Helm Secrets Version: `v4.2.2`
- ArgoCD Version: `2.4.14`
- OS: `Fedora release 37 (Thirty Seven)`
- Shell: `GNU bash, version 5.2.9(1)-release (x86_64-redhat-linux-gnu)`

### Anything else?

I've fixed this issue by updating [`scripts/lib/file/http.sh`](https://github.com/jkroepke/helm-secrets/blob/main/scripts/lib/file/http.sh#L18).

To remove special `&` and `?` chars messing with bash, I've replaced : 

```
_tmp_file="$(_mktemp).${_url##*.}"
```

by : 

```
_tmp_file="$(_mktemp)"
```

I don't know what are the implications of this change, so I opened this issue to discuss it.

Regards,
Antoine

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 - Performance issues with large value files (vals backend)
+- Remote value file download fails when URL contains query strings
 
 ## [4.2.2] - 2022-11-20
 ### Fixed

--- a/scripts/lib/file/http.sh
+++ b/scripts/lib/file/http.sh
@@ -15,7 +15,7 @@ _file_http_get() {
         _url="${1}"
     fi
 
-    _tmp_file="$(_mktemp).${_url##*.}"
+    _tmp_file="$(_mktemp)"
 
     if ! download "${_url}" >"${_tmp_file}"; then
         if [ "${IGNORE_MISSING_VALUES}" = "true" ]; then

--- a/tests/unit/template.bats
+++ b/tests/unit/template.bats
@@ -1442,6 +1442,34 @@ load '../bats/extensions/bats-file/load'
     assert_success
 }
 
+@test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url" {
+    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
+    VALUES_PATH="${VALUES}"
+
+    create_chart "${TEST_TEMP_DIR}"
+
+    run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values -f "${VALUES_PATH}" 2>&1
+
+    assert_output -e "\[helm-secrets\] Decrypt: .*${VALUES}"
+    assert_output --partial "global_secret: global_bar"
+    assert_output --partial "[helm-secrets] Removed: "
+    assert_success
+}
+
+@test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url + --set-file" {
+    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
+    VALUES_PATH="${VALUES}"
+
+    create_chart "${TEST_TEMP_DIR}"
+
+    run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values --set-file "content=${VALUES_PATH}" 2>&1
+
+    assert_output -e "\[helm-secrets\] Decrypt: .*${VALUES}"
+    assert_output --partial "global_secret: global_bar"
+    assert_output --partial "[helm-secrets] Removed: "
+    assert_success
+}
+
 @test "template: helm template w/ chart + --evaluate-templates" {
     if ! is_backend "vala"; then
         skip

--- a/tests/unit/template.bats
+++ b/tests/unit/template.bats
@@ -1443,6 +1443,10 @@ load '../bats/extensions/bats-file/load'
 }
 
 @test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url" {
+    if on_windows; then
+        skip
+    fi
+
     VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
     VALUES_PATH="${VALUES}"
 
@@ -1450,13 +1454,17 @@ load '../bats/extensions/bats-file/load'
 
     run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values -f "${VALUES_PATH}" 2>&1
 
-    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
+    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
     assert_output --partial "global_secret: global_bar"
     assert_output --partial "[helm-secrets] Removed: "
     assert_success
 }
 
 @test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url + --set-file" {
+    if on_windows; then
+        skip
+    fi
+
     VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
     VALUES_PATH="${VALUES}"
 
@@ -1464,7 +1472,7 @@ load '../bats/extensions/bats-file/load'
 
     run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values --set-file "content=${VALUES_PATH}" 2>&1
 
-    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
+    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
     assert_output --partial "global_secret: global_bar"
     assert_output --partial "[helm-secrets] Removed: "
     assert_success

--- a/tests/unit/template.bats
+++ b/tests/unit/template.bats
@@ -1443,7 +1443,7 @@ load '../bats/extensions/bats-file/load'
 }
 
 @test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url" {
-    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
+    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
     VALUES_PATH="${VALUES}"
 
     create_chart "${TEST_TEMP_DIR}"
@@ -1457,7 +1457,7 @@ load '../bats/extensions/bats-file/load'
 }
 
 @test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url + --set-file" {
-    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
+    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
     VALUES_PATH="${VALUES}"
 
     create_chart "${TEST_TEMP_DIR}"

--- a/tests/unit/template.bats
+++ b/tests/unit/template.bats
@@ -1443,28 +1443,28 @@ load '../bats/extensions/bats-file/load'
 }
 
 @test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url" {
-    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
+    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
     VALUES_PATH="${VALUES}"
 
     create_chart "${TEST_TEMP_DIR}"
 
     run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values -f "${VALUES_PATH}" 2>&1
 
-    assert_output -e "\[helm-secrets\] Decrypt: .*${VALUES}"
+    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
     assert_output --partial "global_secret: global_bar"
     assert_output --partial "[helm-secrets] Removed: "
     assert_success
 }
 
 @test "template: helm template w/ remote chart + secrets.yaml + http:// + query string url + --set-file" {
-    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
+    VALUES="https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
     VALUES_PATH="${VALUES}"
 
     create_chart "${TEST_TEMP_DIR}"
 
     run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values --set-file "content=${VALUES_PATH}" 2>&1
 
-    assert_output -e "\[helm-secrets\] Decrypt: .*${VALUES}"
+    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml\?foo=b4r&test=true&arg1="
     assert_output --partial "global_secret: global_bar"
     assert_output --partial "[helm-secrets] Removed: "
     assert_success

--- a/tests/unit/template.bats
+++ b/tests/unit/template.bats
@@ -1454,7 +1454,7 @@ load '../bats/extensions/bats-file/load'
 
     run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values -f "${VALUES_PATH}" 2>&1
 
-    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
+    assert_output -e "\[helm-secrets\] Decrypt: .*${VALUES//\?/\\?}"
     assert_output --partial "global_secret: global_bar"
     assert_output --partial "[helm-secrets] Removed: "
     assert_success
@@ -1472,7 +1472,7 @@ load '../bats/extensions/bats-file/load'
 
     run "${HELM_BIN}" secrets template --repo https://jkroepke.github.io/helm-charts/ --version 1.0.3 values --set-file "content=${VALUES_PATH}" 2>&1
 
-    assert_output -e "\[helm-secrets\] Decrypt: .*https://raw.githubusercontent.com/jkroepke/helm-secrets/main/tests/assets/values/${HELM_SECRETS_BACKEND}/secrets.yaml?foo=b4r&test=true&arg1="
+    assert_output -e "\[helm-secrets\] Decrypt: .*${VALUES//\?/\\?}"
     assert_output --partial "global_secret: global_bar"
     assert_output --partial "[helm-secrets] Removed: "
     assert_success

